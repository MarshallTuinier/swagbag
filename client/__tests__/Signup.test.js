import { mount } from "enzyme";
import wait from "waait";
import { ApolloConsumer } from "react-apollo";
import { MockedProvider } from "react-apollo/test-utils";
import toJSON from "enzyme-to-json";
import { SIGNUP_MUTATION, CURRENT_USER_QUERY } from "../GraphQL";
import SignupForm from "../components/SignupForm";
import { fakeUser } from "../lib/testUtils";

// Helper function to simulate typing in a form
const type = (wrapper, name, value) => {
  wrapper.find(`input[name="${name}"]`).simulate("change", {
    target: { name, value }
  });
};

const user = fakeUser();

const mocks = [
  // Signup mutation mock
  {
    request: {
      query: SIGNUP_MUTATION,
      variables: {
        email: user.email,
        name: user.name,
        password: "test",
        confirmPassword: "test"
      }
    },
    result: {
      data: {
        signup: {
          __typename: "User",
          id: "abc123",
          email: user.email,
          name: user.name
        }
      }
    }
  },
  // Current user mock
  {
    request: { query: CURRENT_USER_QUERY },
    result: {
      data: {
        currentUser: {
          ...user
        }
      }
    }
  }
];

describe("<SignupForm />", async () => {
  it("renders and matches snapshot", async () => {
    const wrapper = mount(
      <MockedProvider>
        <SignupForm />
      </MockedProvider>
    );
    expect(toJSON(wrapper.find("form"))).toMatchSnapshot();
  });

  it("calls the mutation properly", async () => {
    let apolloClient;
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <ApolloConsumer>
          {client => {
            apolloClient = client;
            return <SignupForm />;
          }}
        </ApolloConsumer>
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    type(wrapper, "name", user.name);
    type(wrapper, "email", user.email);
    type(wrapper, "password", "test");
    type(wrapper, "confirmPassword", "test");
    wrapper.update();
    wrapper.find("form").simulate("submit");
    await wait();
    // Query the user out of the apollo client
    const currentUser = await apolloClient.query({ query: CURRENT_USER_QUERY });
    expect(currentUser.data.currentUser).toMatchObject(user);
  });
});
