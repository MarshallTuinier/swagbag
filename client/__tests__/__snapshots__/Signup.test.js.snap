// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SignupForm /> renders and matches snapshot 1`] = `
<form
  className="SignupForm__Form-cn2ap0-0 xJdyd"
  method="post"
  onSubmit={[Function]}
>
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <h2>
      Sign Up for an Account
    </h2>
    <DisplayError
      error={Object {}}
    />
    <label
      htmlFor="email"
    >
      Email
    </label>
    <input
      name="email"
      onChange={[Function]}
      placeholder="Email"
      required={true}
      type="email"
      value=""
    />
    <label
      htmlFor="name"
    >
      Name
    </label>
    <input
      name="name"
      onChange={[Function]}
      placeholder="Name"
      required={true}
      type="text"
      value=""
    />
    <label
      htmlFor="password"
    >
      Password
    </label>
    <input
      name="password"
      onChange={[Function]}
      placeholder="Password"
      required={true}
      type="password"
      value=""
    />
    <label
      htmlFor="confirmPassword"
    >
      Confirm Password
    </label>
    <input
      name="confirmPassword"
      onChange={[Function]}
      placeholder="Confirm Password"
      required={true}
      type="password"
      value=""
    />
    <button
      type="submit"
    >
      Sign Up!
    </button>
  </fieldset>
</form>
`;
